### PR TITLE
Phase4

### PR DESCRIPTION
**Pull request by hu55a1n1**
Implements Phase 4 of #1 - Add support for IBC Connection machinery

- [x] complete/import the ics03 machinery from ibc-rs
- [x] state will now need to be able to store connection ends
- [x] should now be able to manually complete connection handshake between two chains

### Testing
1. Follow the steps outlined in the README.md to setup a test environment running a basecoin chain (i.e. `basecoin-0`) and two ibc chains (`ibc-0` & `ibc-1`). 
2. Run the following command to create a connection between `basecoin-0` and `ibc-0` -> 
```sh
$ RUST_BACKTRACE=1 hermes create connection ibc-0 basecoin-0
# ...
Success: Connection {
    delay_period: 0ns,
    a_side: ConnectionSide {
        chain: ProdChainHandle {
            chain_id: ChainId {
                id: "ibc-0",
                version: 0,
            },
            runtime_sender: Sender { .. },
        },
        client_id: ClientId(
            "07-tendermint-0",
        ),
        connection_id: Some(
            ConnectionId(
                "connection-0",
            ),
        ),
    },
    b_side: ConnectionSide {
        chain: ProdChainHandle {
            chain_id: ChainId {
                id: "basecoin-0",
                version: 0,
            },
            runtime_sender: Sender { .. },
        },
        client_id: ClientId(
            "07-tendermint-0",
        ),
        connection_id: Some(
            ConnectionId(
                "connection-0",
            ),
        ),
    },
}
```

### CODE REVIEW

**Comment by hu55a1n1** [comment]
Note that opening a connection in the opposite direction may fail intermittently - this is a known bug in Tendermock's AVL impl. See the tracking issue here -> https://github.com/CharlyCst/tendermock/issues/2

**Comment by hu55a1n1** [comment]
Also note that we must depend on the `basecoin/phase-4` branch of ibc-rs for multiple reasons which include ->
* Verification methods in the ibc-rs module are incomplete and call `todo!()`. 
* basecoin-rs uses a different `ProofSpec` and ibc-rs currently assumes that all Tendermint chains use the Cosmos SDK specific `ProofSpec`.

**Comment by thanethomson** [comment]
Sorry for the lag in getting to this. I've been trying to set this up today and I haven't been successful in creating a connection unfortunately.

See [this gist](https://gist.github.com/thanethomson/9b0fdf14ff00037840d9467a6eae5237) for my output logs.

When you say this:

> Also note that we must depend on the basecoin/phase-4 branch of ibc-rs for multiple reasons

Does that mean that I need to build `hermes` from that branch?

**Comment by hu55a1n1** [comment]
> Does that mean that I need to build hermes from that branch?

That comment is outdated, sorry for not updating it. We need to build hermes from the `basecoin/phase-4-1` branch, because we need a version of hermes that has the same proto defs and support for custom `ProofSpecs`. We also need to tell hermes to use basecoin's `ProofSpecs` in the `config.toml`. 
I'll update the README.md with this info. 
```toml
[[chains]]
id = 'basecoin-0'
rpc_addr = 'http://127.0.0.1:26357'
grpc_addr = 'http://127.0.0.1:9093'
websocket_addr = 'ws://localhost:26357/websocket'
rpc_timeout = '10s'
account_prefix = 'cosmos'
key_name = 'testkey'
store_prefix = 'ibc'
gas_price = { price = 0.001, denom = 'stake' }
clock_drift = '5s'
trusting_period = '14days'
proof_specs = '''
[
  {
    "leaf_spec": {
      "hash": 1,
      "prehash_key": 0,
      "prehash_value": 0,
      "length": 0,
      "prefix": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "inner_spec": {
      "child_order": [
        0,
        1,
        2
      ],
      "child_size": 32,
      "min_prefix_length": 0,
      "max_prefix_length": 64,
      "empty_child": [
        0,
        32
      ],
      "hash": 1
    },
    "max_depth": 0,
    "min_depth": 0
  },
  {
    "leaf_spec": {
      "hash": 1,
      "prehash_key": 0,
      "prehash_value": 0,
      "length": 0,
      "prefix": [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
    },
    "inner_spec": {
      "child_order": [
        0,
        1,
        2
      ],
      "child_size": 32,
      "min_prefix_length": 0,
      "max_prefix_length": 64,
      "empty_child": [
        0,
        32
      ],
      "hash": 1
    },
    "max_depth": 0,
    "min_depth": 0
  }
]
'''
```

File: src/app/mod.rs [hunk 1]
```diff
@@ -171,16 +185,44 @@ impl<S: ProvableStore + 'static> Application for BaseCoinApp<S> {
...
+                        proof.encode(&mut buffer).unwrap(); // safety - cannot fail since buf is a vector
+
+                        let mut ops = vec![];
+                        if let Some(mut proofs) = result.proof {
+                            ops.append(&mut proofs);
+                        }
+                        ops.push(ProofOp {
+                            r#type: "".to_string(),
+                            // FIXME(hu55a1n1)
+                            key: "ibc".to_string().into_bytes(),
```
Comment (thanethomson) [inline src/app/mod.rs:210]: What does this `FIXME` mean? i.e. how should it be fixed?

File: src/app/mod.rs [hunk 2]
```diff
@@ -240,9 +282,20 @@ impl<S: ProvableStore + 'static> Application for BaseCoinApp<S> {
     }
 
     fn commit(&self) -> ResponseCommit {
+        let mut modules = self.modules.write().unwrap();
+        for m in modules.iter_mut() {
+            m.commit().expect("failed to commit to state");
```
Comment (thanethomson) [inline src/app/mod.rs:287]: What happens in the case of a partially committed set of modules? For example, if we had 4 modules, where 2 of them committed successfully and then the 3rd one failed (and panicked)?

Will the commits from modules 1 and 2 be rolled back if we restart the application?

**Review by thanethomson — APPROVED** [review]
Overall this looks great @hu55a1n1! :tada: Well done! A lot of complex stuff going on here.

I just had a few questions, but I don't think that should stop you from merging this. If my questions should be addressed in the code we can create follow-up issues.

File: src/app/mod.rs [hunk 3]
```diff
@@ -47,48 +53,56 @@ use tendermint_proto::abci::{
...
         // `SubStore` guarantees modules exclusive access to all paths in the store key-space.
-        let modules: Vec<Box<dyn Module + Send + Sync>> = vec![
-            Box::new(Bank {
-                store: SubStore::new(store.clone(), prefix::Bank),
-            }),
-            Box::new(Ibc {
-                store: SubStore::new(store.clone(), prefix::Ibc),
-                client_counter: 0,
-            }),
+        let modules: Vec<Box<dyn Module<ModuleStore<S>> + Send + Sync>> = vec![
```
[the remark this answers is not in the document]
Comment (hu55a1n1) [reply src/app/mod.rs:81]: Good question! I've been thinking about this myself recently. 🤔 One possible alternative would be to have an `enum` with one variant per module, but that would imply that we know of all supported modules beforehand and that is not desirable if we think of this code as part of a framework. 
Furthermore, ICS25 requires that -> 
> The module set should be dynamic: chains should be able to add and destroy modules, which can themselves bind to and unbind from ports, at will with a persistent IBC handler.

So I am not aware of a better alternative but happy to discuss other ideas/solutions.

File: src/app/mod.rs (hunk 1 above)
Comment (hu55a1n1) [reply src/app/mod.rs:210]: Nice catch! We must return the module's prefix here as the key instead of hard-coding `ibc`. I'll create an issue for this.

Comment (hu55a1n1) [reply src/app/mod.rs:210]: https://github.com/informalsystems/basecoin-rs/issues/31

File: src/app/mod.rs (hunk 2 above)
Comment (hu55a1n1) [reply src/app/mod.rs:287]: basecoin-rs doesn't provide any exception safety guarantees as of now. The reason for this is that we only provide a volatile in-memory data store currently. But this is definitely something we should address in the future. 
I do have some ideas for how to achieve this with minimal effort using rust's `panic_handler` to write a crash log containing pending writes. I'll try to document the idea in an issue.

Comment (hu55a1n1) [reply src/app/mod.rs:287]: https://github.com/informalsystems/basecoin-rs/issues/32

**Pull request merged**